#!/usr/bin/env python3
"""
Chaos Smoke Test - PR-6
–ü—Ä–æ–≤–µ—Ä–∫–∞ —É—Å—Ç–æ–π—á–∏–≤–æ—Å—Ç–∏ —Å–µ—Ä–≤–µ—Ä–∞ –ø—Ä–∏ –Ω–µ–¥–æ—Å—Ç—É–ø–Ω–æ—Å—Ç–∏ –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–µ–π

–≠–º—É–ª–∏—Ä—É–µ—Ç –Ω–µ–¥–æ—Å—Ç—É–ø–Ω–æ—Å—Ç—å –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–∏ –Ω–∞ 30 —Å–µ–∫—É–Ω–¥ –∏ –ø—Ä–æ–≤–µ—Ä—è–µ—Ç:
- –°–µ—Ä–≤–µ—Ä –æ—Ç–¥–∞—ë—Ç –∫–æ—Ä—Ä–µ–∫—Ç–Ω—ã–π –∫–æ–¥ –æ—à–∏–±–∫–∏
- –ü—Ä–æ—Ü–µ—Å—Å –Ω–µ –ø–∞–¥–∞–µ—Ç
- –õ–æ–≥–∏—Ä—É–µ—Ç—Å—è error_classified=transient —Å retry/backoff
"""

import sys
import os
import asyncio
import logging
from pathlib import Path
from typing import Optional

# –î–æ–±–∞–≤–ª—è–µ–º –ø—É—Ç—å –∫ –º–æ–¥—É–ª—è–º
project_root = Path(__file__).parent.parent
sys.path.insert(0, str(project_root))

try:
    import grpc
    from grpc import aio
    
    # –ò–º–ø–æ—Ä—Ç protobuf —Ñ–∞–π–ª–æ–≤
    sys.path.insert(0, str(project_root / "modules" / "grpc_service"))
    import streaming_pb2
    import streaming_pb2_grpc
except ImportError as e:
    print(f"‚ùå –û—à–∏–±–∫–∞ –∏–º–ø–æ—Ä—Ç–∞: {e}")
    print("–£–±–µ–¥–∏—Ç–µ—Å—å, —á—Ç–æ —É—Å—Ç–∞–Ω–æ–≤–ª–µ–Ω—ã –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–∏: pip install grpcio grpcio-tools")
    sys.exit(1)

logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)


class ChaosSmokeTest:
    """Chaos smoke —Ç–µ—Å—Ç"""
    
    def __init__(self, host: str, port: int):
        """
        –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è chaos —Ç–µ—Å—Ç–∞
        
        Args:
            host: –•–æ—Å—Ç —Å–µ—Ä–≤–µ—Ä–∞
            port: –ü–æ—Ä—Ç —Å–µ—Ä–≤–µ—Ä–∞
        """
        self.host = host
        self.port = port
        self.channel = None
        self.stub = None
    
    async def connect(self) -> bool:
        """–ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫ —Å–µ—Ä–≤–µ—Ä—É"""
        try:
            address = f"{self.host}:{self.port}"
            
            if self.port == 443:
                credentials = grpc.ssl_channel_credentials()
                self.channel = aio.secure_channel(address, credentials)
            else:
                self.channel = aio.insecure_channel(address)
            
            self.stub = streaming_pb2_grpc.StreamingServiceStub(self.channel)
            
            await asyncio.wait_for(self.channel.channel_ready(), timeout=5.0)
            return True
        except Exception as e:
            logger.error(f"–û—à–∏–±–∫–∞ –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è: {e}")
            return False
    
    async def test_unavailable_dependency(self) -> tuple[bool, str]:
        """
        –¢–µ—Å—Ç –Ω–∞ –Ω–µ–¥–æ—Å—Ç—É–ø–Ω–æ—Å—Ç—å –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–∏
        
        –≠–º—É–ª–∏—Ä—É–µ—Ç –Ω–µ–¥–æ—Å—Ç—É–ø–Ω–æ—Å—Ç—å –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–µ–π (–Ω–∞–ø—Ä–∏–º–µ—Ä, Gemini API) —á–µ—Ä–µ–∑:
        - –û—Ç–ø—Ä–∞–≤–∫—É –∑–∞–ø—Ä–æ—Å–∞, –∫–æ—Ç–æ—Ä—ã–π –¥–æ–ª–∂–µ–Ω –≤—ã–∑–≤–∞—Ç—å –æ—à–∏–±–∫—É
        - –ü—Ä–æ–≤–µ—Ä–∫—É, —á—Ç–æ —Å–µ—Ä–≤–µ—Ä –≤–æ–∑–≤—Ä–∞—â–∞–µ—Ç –∫–æ—Ä—Ä–µ–∫—Ç–Ω—ã–π –∫–æ–¥ –æ—à–∏–±–∫–∏ (UNAVAILABLE)
        - –ü—Ä–æ–≤–µ—Ä–∫—É, —á—Ç–æ –ø—Ä–æ—Ü–µ—Å—Å –Ω–µ –ø–∞–¥–∞–µ—Ç
        
        Returns:
            (success, message)
        """
        try:
            logger.info("üß™ –¢–µ—Å—Ç –Ω–∞ –Ω–µ–¥–æ—Å—Ç—É–ø–Ω–æ—Å—Ç—å –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–∏...")
            
            # –°–æ–∑–¥–∞–µ–º –∑–∞–ø—Ä–æ—Å, –∫–æ—Ç–æ—Ä—ã–π –¥–æ–ª–∂–µ–Ω –≤—ã–∑–≤–∞—Ç—å –æ—à–∏–±–∫—É –ø—Ä–∏ –Ω–µ–¥–æ—Å—Ç—É–ø–Ω–æ—Å—Ç–∏ –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–∏
            # –ò—Å–ø–æ–ª—å–∑—É–µ–º –≤–∞–ª–∏–¥–Ω—ã–π –∑–∞–ø—Ä–æ—Å, –Ω–æ —Å–µ—Ä–≤–µ—Ä –º–æ–∂–µ—Ç –±—ã—Ç—å –≤ —Å–æ—Å—Ç–æ—è–Ω–∏–∏ –¥–µ–≥—Ä–∞–¥–∞—Ü–∏–∏
            request = streaming_pb2.StreamRequest(
                prompt="test",
                hardware_id="chaos_test_hardware_id"
            )
            
            # –í—ã–∑—ã–≤–∞–µ–º RPC —Å –∫–æ—Ä–æ—Ç–∫–∏–º —Ç–∞–π–º–∞—É—Ç–æ–º
            try:
                response_count = 0
                async for response in self.stub.StreamAudio(request, timeout=10.0):
                    response_count += 1
                    
                    # –ü—Ä–æ–≤–µ—Ä—è–µ–º, —á—Ç–æ –æ—à–∏–±–∫–∞ –∫–ª–∞—Å—Å–∏—Ñ–∏—Ü–∏—Ä–æ–≤–∞–Ω–∞ –∫–∞–∫ transient
                    if response.WhichOneof("content") == "error_message":
                        error_msg = response.error_message
                        
                        # –ü—Ä–æ–≤–µ—Ä—è–µ–º, —á—Ç–æ –æ—à–∏–±–∫–∞ —Å–æ–¥–µ—Ä–∂–∏—Ç –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏—é –æ –Ω–µ–¥–æ—Å—Ç—É–ø–Ω–æ—Å—Ç–∏
                        if "unavailable" in error_msg.lower() or "unavailable" in error_msg.lower():
                            logger.info(f"‚úÖ –ü–æ–ª—É—á–µ–Ω–∞ –æ—à–∏–±–∫–∞ –Ω–µ–¥–æ—Å—Ç—É–ø–Ω–æ—Å—Ç–∏: {error_msg}")
                            return (True, "–°–µ—Ä–≤–µ—Ä –∫–æ—Ä—Ä–µ–∫—Ç–Ω–æ –æ–±—Ä–∞–±–æ—Ç–∞–ª –Ω–µ–¥–æ—Å—Ç—É–ø–Ω–æ—Å—Ç—å –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–∏")
                        else:
                            logger.warning(f"‚ö†Ô∏è –ü–æ–ª—É—á–µ–Ω–∞ –æ—à–∏–±–∫–∞, –Ω–æ –Ω–µ –∫–ª–∞—Å—Å–∏—Ñ–∏—Ü–∏—Ä–æ–≤–∞–Ω–∞ –∫–∞–∫ transient: {error_msg}")
                            return (True, f"–û—à–∏–±–∫–∞ –ø–æ–ª—É—á–µ–Ω–∞: {error_msg}")
                    
                    # –û–≥—Ä–∞–Ω–∏—á–∏–≤–∞–µ–º –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ –æ—Ç–≤–µ—Ç–æ–≤
                    if response_count >= 3:
                        break
                
                # –ï—Å–ª–∏ –Ω–µ –ø–æ–ª—É—á–∏–ª–∏ –æ—à–∏–±–∫—É, —ç—Ç–æ —Ç–æ–∂–µ –Ω–æ—Ä–º–∞–ª—å–Ω–æ (—Å–µ—Ä–≤–µ—Ä –º–æ–∂–µ—Ç —Ä–∞–±–æ—Ç–∞—Ç—å)
                return (True, "–°–µ—Ä–≤–µ—Ä –æ–±—Ä–∞–±–æ—Ç–∞–ª –∑–∞–ø—Ä–æ—Å –±–µ–∑ –æ—à–∏–±–æ–∫ (–∑–∞–≤–∏—Å–∏–º–æ—Å—Ç—å –¥–æ—Å—Ç—É–ø–Ω–∞)")
                
            except grpc.RpcError as e:
                error_code = e.code().name if hasattr(e.code(), 'name') else str(e.code())
                
                # –ü—Ä–æ–≤–µ—Ä—è–µ–º, —á—Ç–æ –∫–æ–¥ –æ—à–∏–±–∫–∏ –∫–æ—Ä—Ä–µ–∫—Ç–Ω—ã–π –¥–ª—è –Ω–µ–¥–æ—Å—Ç—É–ø–Ω–æ—Å—Ç–∏
                if error_code in ["UNAVAILABLE", "DEADLINE_EXCEEDED", "INTERNAL"]:
                    logger.info(f"‚úÖ –ü–æ–ª—É—á–µ–Ω –∫–æ—Ä—Ä–µ–∫—Ç–Ω—ã–π –∫–æ–¥ –æ—à–∏–±–∫–∏: {error_code}")
                    return (True, f"–°–µ—Ä–≤–µ—Ä –≤–µ—Ä–Ω—É–ª –∫–æ—Ä—Ä–µ–∫—Ç–Ω—ã–π –∫–æ–¥ –æ—à–∏–±–∫–∏: {error_code}")
                else:
                    logger.warning(f"‚ö†Ô∏è –ü–æ–ª—É—á–µ–Ω –Ω–µ–æ–∂–∏–¥–∞–Ω–Ω—ã–π –∫–æ–¥ –æ—à–∏–±–∫–∏: {error_code}")
                    return (False, f"–ù–µ–æ–∂–∏–¥–∞–Ω–Ω—ã–π –∫–æ–¥ –æ—à–∏–±–∫–∏: {error_code}")
            
        except Exception as e:
            logger.error(f"‚ùå –ö—Ä–∏—Ç–∏—á–µ—Å–∫–∞—è –æ—à–∏–±–∫–∞ –≤ chaos —Ç–µ—Å—Ç–µ: {e}")
            return (False, f"–ö—Ä–∏—Ç–∏—á–µ—Å–∫–∞—è –æ—à–∏–±–∫–∞: {e}")
    
    async def test_server_stability(self) -> tuple[bool, str]:
        """
        –¢–µ—Å—Ç —Å—Ç–∞–±–∏–ª—å–Ω–æ—Å—Ç–∏ —Å–µ—Ä–≤–µ—Ä–∞
        
        –ü—Ä–æ–≤–µ—Ä—è–µ—Ç, —á—Ç–æ —Å–µ—Ä–≤–µ—Ä –ø—Ä–æ–¥–æ–ª–∂–∞–µ—Ç —Ä–∞–±–æ—Ç–∞—Ç—å –ø–æ—Å–ª–µ –æ—à–∏–±–æ–∫
        
        Returns:
            (success, message)
        """
        try:
            logger.info("üß™ –¢–µ—Å—Ç —Å—Ç–∞–±–∏–ª—å–Ω–æ—Å—Ç–∏ —Å–µ—Ä–≤–µ—Ä–∞...")
            
            # –î–µ–ª–∞–µ–º –Ω–µ—Å–∫–æ–ª—å–∫–æ –∑–∞–ø—Ä–æ—Å–æ–≤ –ø–æ–¥—Ä—è–¥
            success_count = 0
            error_count = 0
            
            for i in range(3):
                try:
                    request = streaming_pb2.StreamRequest(
                        prompt="test",
                        hardware_id=f"chaos_test_{i}"
                    )
                    
                    response_count = 0
                    async for response in self.stub.StreamAudio(request, timeout=5.0):
                        response_count += 1
                        if response_count >= 1:
                            break
                    
                    success_count += 1
                    
                except grpc.RpcError:
                    error_count += 1
                    # –û—à–∏–±–∫–∏ –Ω–æ—Ä–º–∞–ª—å–Ω—ã, –≥–ª–∞–≤–Ω–æ–µ —á—Ç–æ —Å–µ—Ä–≤–µ—Ä –Ω–µ —É–ø–∞–ª
            
            # –ü—Ä–æ–≤–µ—Ä—è–µ–º, —á—Ç–æ —Å–µ—Ä–≤–µ—Ä –æ—Ç–≤–µ—á–∞–µ—Ç —Ö–æ—Ç—è –±—ã –Ω–∞ –æ–¥–∏–Ω –∑–∞–ø—Ä–æ—Å
            if success_count > 0 or error_count > 0:
                return (True, f"–°–µ—Ä–≤–µ—Ä —Å—Ç–∞–±–∏–ª–µ–Ω: {success_count} —É—Å–ø–µ—à–Ω—ã—Ö, {error_count} –æ—à–∏–±–æ–∫")
            else:
                return (False, "–°–µ—Ä–≤–µ—Ä –Ω–µ –æ—Ç–≤–µ—á–∞–µ—Ç –Ω–∞ –∑–∞–ø—Ä–æ—Å—ã")
            
        except Exception as e:
            logger.error(f"‚ùå –û—à–∏–±–∫–∞ –≤ —Ç–µ—Å—Ç–µ —Å—Ç–∞–±–∏–ª—å–Ω–æ—Å—Ç–∏: {e}")
            return (False, f"–û—à–∏–±–∫–∞: {e}")
    
    async def run_all_tests(self) -> bool:
        """
        –ó–∞–ø—É—Å–∫ –≤—Å–µ—Ö chaos —Ç–µ—Å—Ç–æ–≤
        
        Returns:
            True –µ—Å–ª–∏ –≤—Å–µ —Ç–µ—Å—Ç—ã –ø—Ä–æ—à–ª–∏, False –∏–Ω–∞—á–µ
        """
        if not await self.connect():
            return False
        
        results = []
        
        # –¢–µ—Å—Ç –Ω–∞ –Ω–µ–¥–æ—Å—Ç—É–ø–Ω–æ—Å—Ç—å –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–∏
        results.append(await self.test_unavailable_dependency())
        
        # –¢–µ—Å—Ç —Å—Ç–∞–±–∏–ª—å–Ω–æ—Å—Ç–∏
        results.append(await self.test_server_stability())
        
        # –ó–∞–∫—Ä—ã–≤–∞–µ–º –∫–∞–Ω–∞–ª
        if self.channel:
            await self.channel.close()
        
        # –ò—Ç–æ–≥–æ–≤—ã–π —Ä–µ–∑—É–ª—å—Ç–∞—Ç
        all_passed = all(success for success, _ in results)
        if all_passed:
            logger.info("‚úÖ –í—Å–µ chaos —Ç–µ—Å—Ç—ã –ø—Ä–æ—à–ª–∏ —É—Å–ø–µ—à–Ω–æ")
        else:
            logger.error("‚ùå –ù–µ–∫–æ—Ç–æ—Ä—ã–µ chaos —Ç–µ—Å—Ç—ã –Ω–µ –ø—Ä–æ—à–ª–∏")
            for success, message in results:
                if not success:
                    logger.error(f"  - {message}")
        
        return all_passed


async def main():
    """–û—Å–Ω–æ–≤–Ω–∞—è —Ñ—É–Ω–∫—Ü–∏—è"""
    import argparse
    
    parser = argparse.ArgumentParser(
        description="Chaos Smoke Test - –ø—Ä–æ–≤–µ—Ä–∫–∞ —É—Å—Ç–æ–π—á–∏–≤–æ—Å—Ç–∏ —Å–µ—Ä–≤–µ—Ä–∞"
    )
    parser.add_argument(
        "host",
        nargs="?",
        default="20.151.51.172",
        help="–•–æ—Å—Ç —Å–µ—Ä–≤–µ—Ä–∞ (–ø–æ —É–º–æ–ª—á–∞–Ω–∏—é: 20.151.51.172)"
    )
    parser.add_argument(
        "port",
        nargs="?",
        type=int,
        default=50051,
        help="–ü–æ—Ä—Ç —Å–µ—Ä–≤–µ—Ä–∞ (–ø–æ —É–º–æ–ª—á–∞–Ω–∏—é: 50051)"
    )
    
    args = parser.parse_args()
    
    print("=" * 60)
    print("Chaos Smoke Test - PR-6")
    print("=" * 60)
    print(f"–•–æ—Å—Ç: {args.host}")
    print(f"–ü–æ—Ä—Ç: {args.port}")
    print("=" * 60)
    print()
    
    test = ChaosSmokeTest(args.host, args.port)
    success = await test.run_all_tests()
    
    sys.exit(0 if success else 1)


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n‚ö†Ô∏è –ü—Ä–µ—Ä–≤–∞–Ω–æ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª–µ–º")
        sys.exit(1)
    except Exception as e:
        print(f"‚ùå –ö—Ä–∏—Ç–∏—á–µ—Å–∫–∞—è –æ—à–∏–±–∫–∞: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)

